Process termination test for SGX

mzohreva opened the pull request:
The issue is described in https://github.com/fortanix/rust-sgx/issues/109

cc @jethrogb

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nikomatsakis (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

nikomatsakis commented:
I have no particular problem adding this test to the rust repo, but I'm curious if the Rust repo is the best place for it, since it seems more like a bug in https://github.com/fortanix/rust-sgx/. 

(I'm not sure how things are setup or exactly what the role of the rust-sgx project is, to be clear.)

jethrogb commented on src/test/run-make/sgx-process-termination/Makefile:
```diff
@@ -0,0 +1,8 @@
+-include ../../run-make-fulldeps/tools.mk
+
+# only-sgx
```
As discussed, I think this should run on all platforms?

mzohreva replied:
Some platforms may need a platform-specific way of running the code (see wasm-* or thumb-* tests in run-make for example), do you want to have platform-specific code here to run the test for those?

jethrogb replied:
I think we should just add `ignore` statements for platforms that have special requirements. Everything that can just run with `cargo run` should get tested.

mzohreva commented:
It is true that there is a bug in rust-sgx at the moment, but it is possible to inadvertently create a bug in rust itself w.r.t this test. We think it is better to include this test here, so we can catch both kinds.

mzohreva commented on src/test/run-make/sgx-process-termination/Makefile:
```diff
@@ -0,0 +1,8 @@
+all:
+	$(RUSTC) accept.rs --target x86_64-fortanix-unknown-sgx
+	$(FTXSGX_ELF2SGXS) $(TMPDIR)/accept --heap-size 0x1001000 --stack-size 0x20000 --threads 10
+	timeout 10s $(FTXSGX_RUNNER) $(TMPDIR)/accept.sgxs
```
I looked at the other tests in run-make and they almost all use `rustc` directly with some machinery in `run-make-fulldeps/tools.mk`. The 2 exceptions using cargo that I found looked hacky:
- https://github.com/rust-lang/rust/blob/master/src/test/run-make/thumb-none-cortex-m/Makefile
- https://github.com/rust-lang/rust/blob/master/src/test/run-make/thumb-none-qemu/script.sh

jethrogb replied:
I think those hacks (i.e. setting `RUSTC_BOOTSTRAP=1`) may be removed, as I don't see any nightly features set in [Cargo.toml](https://github.com/rust-lang/rust/blob/master/Cargo.toml)

mzohreva commented on src/test/run-make/sgx-process-termination/accept.rs:
```diff
@@ -0,0 +1,15 @@
+    thread::spawn(move || {
+        let listen = TcpListener::bind("0:0").unwrap();
+        tx.send(()).unwrap();
+        while let Ok(_) = listen.accept() {}
```
True, but I think there is no reasonable way to avoid that race: we need one thread to be blocked on some I/O and another thread to cause process exit. While the first thread is blocked on I/O it cannot notify the other thread (by definition), and the second thread does not have a way of knowing the first thread is truly blocked on I/O on its own unless we are willing to check `/proc/self/task/*/stat` or something similar.

I think we can reasonably expect that the main thread does not exit before the spawned thread blocks on `accept()` in most cases.

jethrogb replied:
Maybe you can do `thread::yield_now` a bunch after `recv`

mzohreva replied:
I think the `println!` is sufficient, but `thread::yield_now` maybe more readable.

nikomatsakis commented:
OK, let me know when y'all feel this is ready for review.

jethrogb commented:
LGTM

mzohreva commented:
I squashed the commits, it's ready for review

nikomatsakis commented on src/test/run-make/process-termination/Makefile:
```diff
@@ -0,0 +1,20 @@
+	cp -a $(HERE)/$(CRATE) $(WORK_DIR)/$(CRATE)
+	cd $(WORK_DIR)/$(CRATE) && \
+	$(CARGO) build --target $(TARGET) --verbose && \
+	timeout 10s $(CARGO) run --target $(TARGET)
```
So why exactly does this need to be a run-make test -- is it this `timeout 10s` call? i.e., why can't we just make this a normal `run-pass` test?

joelpalmer commented:
Ping from Triage:  Hi @mzohreva, any updates?

mzohreva commented:
sorry for the delay, I rebased and restructured the test according to our previous discussion. It is now in `src/test/ui/process-termination`. I also added another test which is in the same category (and has already found a bug in our rust-sgx implementation)

nikomatsakis commented:
@bors r+ rollup

Thanks!

bors commented:
:pushpin: Commit db1fbd4a11db579436f68be3a8f5fe03484aa45d has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis db1fbd4a11db579436f68be3a8f5fe03484aa45d -->
<!-- homu: {"type":"Approved","sha":"db1fbd4a11db579436f68be3a8f5fe03484aa45d","approver":"nikomatsakis"} -->

Dylan-DPC commented:
fails in rollup

@bors r-

jethrogb commented:
These tests should not be run on the wasm platform, since there is no threading

jethrogb commented:
How come CI breaks on this? Wasm is tier 2 so it shouldn't be running these tests.

Dylan-DPC commented:
investigating

nikomatsakis commented:
Regardless, we should fix it. I seem to recall that wasm might be like tier 1.5. =)

jethrogb commented:
@rustbot modify labels: -S-waiting-on-author +S-waiting-on-review

Dylan-DPC commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 9d8f1170b1fa923373766e26a2847390e0a2ea91 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 9d8f1170b1fa923373766e26a2847390e0a2ea91 -->
<!-- homu: {"type":"Approved","sha":"9d8f1170b1fa923373766e26a2847390e0a2ea91","approver":"nikomatsakis"} -->

Dylan-DPC commented:
failed again in rollup
@bors r- rollup=never

mzohreva commented:
I thought `// ignore-wasm` was sufficient to prevent the tests from running in all wasm targets, but apparently not?

mzohreva commented:
@rustbot modify labels: -S-waiting-on-author +S-waiting-on-review

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 2e749a5ecd3831f8cb947761b15b2b108ca0ec7c has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 2e749a5ecd3831f8cb947761b15b2b108ca0ec7c -->
<!-- homu: {"type":"Approved","sha":"2e749a5ecd3831f8cb947761b15b2b108ca0ec7c","approver":"nikomatsakis"} -->

bors commented:
:hourglass: Testing commit 2e749a5ecd3831f8cb947761b15b2b108ca0ec7c with merge 26604521fcbc365ab54a3966a4cd0254d6de7a0f...
<!-- homu: {"type":"BuildStarted","head_sha":"2e749a5ecd3831f8cb947761b15b2b108ca0ec7c","merge_sha":"26604521fcbc365ab54a3966a4cd0254d6de7a0f"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=28417)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=28417","builder_name":"checks-azure"} -->

mzohreva commented:
The failure seems unrelated to the PR. It [failed during an LLVM build](https://dev.azure.com/rust-lang/rust/_build/results?buildId=28417&view=logs&j=8459d2f4-1fab-5c91-14f9-f1f91c476a8f&t=614b60a3-f1e9-5536-2693-cb0ac1e3b1f4)

nikomatsakis commented:
@bors retry

bors commented:
:hourglass: Testing commit 2e749a5ecd3831f8cb947761b15b2b108ca0ec7c with merge 4249dae4d0c22b8ef962ef97829c6c315a31d3a3...
<!-- homu: {"type":"BuildStarted","head_sha":"2e749a5ecd3831f8cb947761b15b2b108ca0ec7c","merge_sha":"4249dae4d0c22b8ef962ef97829c6c315a31d3a3"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/651940668)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/651940668","builder_name":"checks-actions"} -->

jethrogb commented on src/test/ui/process-termination/process-termination-blocking-io.rs:
```diff
@@ -0,0 +1,19 @@
+fn main() {
+    let (tx, rx) = mpsc::channel();
+    thread::spawn(move || {
+        let listen = TcpListener::bind("0:0").unwrap();
```
Should probably specify `0.0.0.0:0` here

mzohreva replied:
Thanks, fixed

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 2b3adc9b9472e1d1e384422ceac783f9921f3b0b has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 2b3adc9b9472e1d1e384422ceac783f9921f3b0b -->
<!-- homu: {"type":"Approved","sha":"2b3adc9b9472e1d1e384422ceac783f9921f3b0b","approver":"nikomatsakis"} -->

bors commented:
:hourglass: Testing commit 2b3adc9b9472e1d1e384422ceac783f9921f3b0b with merge a2e0b48e6eef854521d3199ee9e327aab298f071...
<!-- homu: {"type":"BuildStarted","head_sha":"2b3adc9b9472e1d1e384422ceac783f9921f3b0b","merge_sha":"a2e0b48e6eef854521d3199ee9e327aab298f071"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/669682516), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=29073)
Approved by: nikomatsakis
Pushing a2e0b48e6eef854521d3199ee9e327aab298f071 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikomatsakis","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=29073","checks-actions":"https://github.com/rust-lang-ci/rust/runs/669682516"},"merge_sha":"a2e0b48e6eef854521d3199ee9e327aab298f071"} -->
